code/game/machinery/slotmachine.dm

```diff
@@ -12,90 +12,94 @@
 	var/resultlvl = null
 
 /obj/machinery/slot_machine/attack_hand(mob/user as mob)
+	tgui_interact(user)
+
+/obj/machinery/slot_machine/tgui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = 0, datum/tgui/master_ui = null, datum/tgui_state/state = GLOB.tgui_default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "SlotMachine", name, 350, 200, master_ui, state)
+		ui.open()
+		ui.set_autoupdate(TRUE)
+
+/obj/machinery/slot_machine/tgui_data(mob/user)
+	var/list/data = list()
+	// Get account
 	account = user.get_worn_id_account()
 	if(!account)
 		if(istype(user.get_active_hand(), /obj/item/card/id))
 			account = get_card_account(user.get_active_hand())
 		else
 			account = null
-	ui_interact(user)
 
-/obj/machinery/slot_machine/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1)
-	ui = SSnanoui.try_update_ui(user, src, ui_key, ui, force_open)
-	if(!ui)
-		ui = new(user, src, ui_key, "slotmachine.tmpl", name, 350, 200)
-		ui.open()
-		ui.set_auto_update(1)
-
-/obj/machinery/slot_machine/ui_data(mob/user, ui_key = "main", datum/topic_state/state = GLOB.default_state)
-	var/data[0]
+	// Send data
 	data["working"] = working
 	data["money"] = account ? account.money : null
 	data["plays"] = plays
 	data["result"] = result
 	data["resultlvl"] = resultlvl
 	return data
 
-/obj/machinery/slot_machine/Topic(href, href_list)
+/obj/machinery/slot_machine/tgui_act(action, params)
+	if(..())
+		return
 	add_fingerprint(usr)
 
-	if(href_list["ops"])
-		if(text2num(href_list["ops"]))	// Play
-			if(working)
-				return
-			if(!account || account.money < 10)
-				return
-			if(!account.charge(10, null, "Bet", "Slot Machine", "Slot Machine"))
-				return
-			plays++
-			working = 1
-			icon_state = "slots-on"
-			playsound(src.loc, 'sound/machines/ding.ogg', 50, 1)
-			addtimer(CALLBACK(src, .proc/spin_slots, usr.name), 25)
+	if(action == "spin")
```

Comment:
Here we check for the action, in this case, `"spin"`. In more complicated UIs, this should be a `switch()` statement, but since this UI has one callback its an if statement.